File: go-backend-interview-prep/10_design_patterns_in_go/04_adapter/main.py
"""Adapter pattern -- Python equivalent of the Go example.

Adapters translate third-party APIs to match your internal interface.
"""


# --- Simulated third-party APIs ---

class TwilioClient:
    def create_message(self, phone_number: str, body: str, from_num: str) -> None:
        print(f"  [TWILIO API] from={from_num} to={phone_number} body={body}")


class SendGridClient:
    def post_mail(self, recipient: str, subject: str, html_body: str) -> None:
        print(f"  [SENDGRID API] to={recipient} subject={subject} body={html_body}")


# --- Adapters: translate to your internal interface ---

class TwilioAdapter:
    def __init__(self, client: TwilioClient, from_num: str):
        self.client = client
        self.from_num = from_num

    def send(self, to: str, message: str) -> None:
        self.client.create_message(to, message, self.from_num)


class SendGridAdapter:
    def __init__(self, client: SendGridClient):
        self.client = client

    def send(self, to: str, message: str) -> None:
        self.client.post_mail(to, "Notification", message)


# --- Service uses YOUR interface ---

class NotificationService:
    def __init__(self, sender):
        self.sender = sender

    def notify(self, to: str, message: str) -> None:
        print(f"  NotificationService.notify -> {to}")
        self.sender.send(to, message)


def main() -> None:
    print("=== Adapter Pattern ===")
    print()

    # Use Twilio via adapter
    print("--- Twilio adapter ---")
    twilio_sender = TwilioAdapter(TwilioClient(), "+1555000000")
    svc1 = NotificationService(twilio_sender)
    svc1.notify("+1234567890", "Your order shipped!")
    print()

    # Use SendGrid via adapter
    print("--- SendGrid adapter ---")
    sg_sender = SendGridAdapter(SendGridClient())
    svc2 = NotificationService(sg_sender)
    svc2.notify("alice@example.com", "Your order shipped!")
    print()

    print("Key: NotificationService depends on .send() duck typing.")
    print("     Adapters translate third-party APIs to match that interface.")
    print("     Swap providers by swapping adapters -- zero service changes.")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/10_design_patterns_in_go/04_adapter/main.go
package main

import "fmt"

// --- Your internal interface ---

type Sender interface {
	Send(to, message string) error
}

// --- Simulated third-party APIs (incompatible signatures) ---

// TwilioClient simulates Twilio's API -- different method names and signatures.
type TwilioClient struct{}

func (t *TwilioClient) CreateMessage(phoneNumber string, body string, from string) error {
	fmt.Printf("  [TWILIO API] from=%s to=%s body=%s\n", from, phoneNumber, body)
	return nil
}

// SendGridClient simulates SendGrid's API -- completely different shape.
type SendGridClient struct{}

func (s *SendGridClient) PostMail(recipient, subject, htmlBody string) error {
	fmt.Printf("  [SENDGRID API] to=%s subject=%s body=%s\n", recipient, subject, htmlBody)
	return nil
}

// --- Adapters: translate third-party APIs to your Sender interface ---

type TwilioAdapter struct {
	client   *TwilioClient
	fromNum  string
}

func (a *TwilioAdapter) Send(to, message string) error {
	// Translate: your simple interface -> Twilio's complex API
	return a.client.CreateMessage(to, message, a.fromNum)
}

type SendGridAdapter struct {
	client *SendGridClient
}

func (a *SendGridAdapter) Send(to, message string) error {
	// Translate: your simple interface -> SendGrid's API
	return a.client.PostMail(to, "Notification", message)
}

// --- Service uses YOUR interface, not third-party types ---

type NotificationService struct {
	sender Sender
}

func (n *NotificationService) Notify(to, message string) {
	fmt.Printf("  NotificationService.Notify -> %s\n", to)
	n.sender.Send(to, message)
}

func main() {
	fmt.Println("=== Adapter Pattern ===")
	fmt.Println()

	// Use Twilio via adapter
	fmt.Println("--- Twilio adapter ---")
	twilioSender := &TwilioAdapter{
		client:  &TwilioClient{},
		fromNum: "+1555000000",
	}
	svc1 := &NotificationService{sender: twilioSender}
	svc1.Notify("+1234567890", "Your order shipped!")
	fmt.Println()

	// Use SendGrid via adapter -- same service, same interface
	fmt.Println("--- SendGrid adapter ---")
	sgSender := &SendGridAdapter{
		client: &SendGridClient{},
	}
	svc2 := &NotificationService{sender: sgSender}
	svc2.Notify("alice@example.com", "Your order shipped!")
	fmt.Println()

	fmt.Println("Key: NotificationService depends on Sender interface.")
	fmt.Println("     Adapters translate third-party APIs to match that interface.")
	fmt.Println("     Swap providers by swapping adapters -- zero service changes.")
}
